lib/galaxy/tools/deps/conda_util.py

```diff
@@ -0,0 +1,305 @@
+        ])
+
+    def env_path(self, env_name):
+        return os.path.join(self.conda_prefix, "envs", env_name)
+
+    def has_env(self, env_name):
+        env_path = self.env_path(env_name)
+        return os.path.isdir(env_path)
+
+    @property
+    def deactivate(self):
+        return self._bin("deactivate")
+
+    @property
+    def activate(self):
+        return self._bin("activate")
+
+    def _bin(self, name):
+        return os.path.join(self.conda_prefix, "bin", name)
+
+
+@six.python_2_unicode_compatible
+class CondaTarget(object):
+
+    def __init__(self, package, version=None, channel=None):
+        if SHELL_UNSAFE_PATTERN.search(package) is not None:
+            raise ValueError("Invalid package [%s] encountered." % package)
+        self.package = package
+        if version and SHELL_UNSAFE_PATTERN.search(version) is not None:
+            raise ValueError("Invalid version [%s] encountered." % version)
+        self.version = version
+        if channel and SHELL_UNSAFE_PATTERN.search(channel) is not None:
+            raise ValueError("Invalid version [%s] encountered." % channel)
+        self.channel = channel
+
+    def __str__(self):
+        attributes = "package=%s" % self.package
+        if self.version is not None:
+            attributes = "%s,version=%s" % (self.package, self.version)
+        else:
+            attributes = "%s,unversioned" % self.package
+
+        if self.channel:
+            attributes = "%s,channel=%s" % self.channel
+
+        return "CondaTarget[%s]" % attributes
+
+    @property
+    def package_specifier(self):
+        """ Return a package specifier as consumed by conda install/create.
+        """
+        if self.version:
+            return "%s=%s" % (self.package, self.version)
+        else:
+            return self.package
+
+    @property
+    def install_environment(self):
+        """ The dependency resolution and installation frameworks will
+        expect each target to be installed it its own environment with
+        a fixed and predictable name given package and version.
+        """
+        if self.version:
+            return "__package__%s@__version__%s" % (self.package, self.version)
+        else:
+            return "__package__%s@__unversion__" % (self.package)
+
+
+def hash_conda_packages(conda_packages, conda_target=None):
+    """ Produce a unique hash on supplied packages.
+    TODO: Ideally we would do this in such a way that preserved environments.
+    """
+    h = hashlib.new('sha256')
+    for conda_package in conda_packages:
+        h.update(conda_package.install_environment)
+    return h.hexdigest()
+
+
+def install_conda(conda_context=None):
+    conda_context = _ensure_conda_context(conda_context)
+    download_cmd = " ".join(commands.download_command(conda_link(), quote_url=True))
+    download_cmd = "%s > /tmp/conda.bash" % download_cmd
+    install_cmd = "bash /tmp/conda.bash -b -p '%s'" % conda_context.conda_prefix
```

Comment:
This one will return an error if the conda_prefix already exists. Depending on the decision about the conda_prefix (Galaxy only or not) we need to add `-f` here.